eng/precommit_local_build.py

```diff
@@ -0,0 +1,99 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+# Python version 3.4 or higher is required to run this script.
+
+# Use case: Runs compilation, testing, and linting for the passed artifacts and POM artifacts.
+#
+# The artifacts must exist in the module list of pom.xml in the root of azure-sdk-for-java. Artifacts may be passed in two ways:
+#
+# 1. Comma separated list of relative POM paths.
+# 2. Comma separated list of groupId:artifactId identifiers.
+#
+# The script must be run at the root of azure-sdk-for-java.
+
+import argparse
+import os
+import xml.etree.ElementTree as ET
+
+baseCommand = 'mvn clean install -f pom.xml -pl "{}" -am {}'
+xmlNamespace = '{http://maven.apache.org/POM/4.0.0}'
+
+def getArtifactsFromPOM(pomPath: str, artifacts: list, debug: bool):
+    if not os.path.exists(pomPath):
+        if debug:
+            print("POM {} doesn't exist, skipping".format(pomPath))
+        return
+
+    tree = ET.parse(pomPath)
+    modulesElement = tree.getroot().find(xmlNamespace + 'modules')
+    if modulesElement != None:
+        pomBasedir = os.path.dirname(pomPath)
+        for modulePomElement in modulesElement.iterfind(xmlNamespace + 'module'):
+            moduleName = modulePomElement.text
+            modulePomPath = os.path.normpath(os.path.join(pomBasedir, moduleName, 'pom.xml'))
+
+            if debug:
+                print('Getting module artifact for {} from aggregator POM {}'.format(moduleName.split('/')[-1], pomPath))
+
+            getArtifactsFromPOM(modulePomPath, artifacts, debug)
+
+    else:
+        groupId = tree.getroot().findtext(xmlNamespace + "groupId")
+        artifactId = tree.getroot().findtext(xmlNamespace + "artifactId")
+        artifactIdentifier = '{}:{}'.format(groupId, artifactId)
+
+        if debug:
+            print('Adding artifact {} for POM file {}'.format(artifactIdentifier, pomPath))
+
+        artifacts.append(artifactIdentifier)
+
+def main():
+    parser = argparse.ArgumentParser(description='Runs compilation, testing, and linting for the passed artifacts.')
+    parser.add_argument('--artifacts', '--a', type=str, default=None, help='Comma separated list of groupId:artifactId identifiers')
+    parser.add_argument('--poms', '--p', type=str, default=None, help='Comma separated list of POM paths')
```

Comment:
Does maven allow passing multiple pom files?

Reply:
This script will "crack" the POM, so if you pass multiple it will grab the `groupId:artifactId` identifier and put it into the `-pl` property. It will also "crack" aggregate POMs into their sub-modules.